quickwit-search/src/client_pool.rs

```diff
@@ -42,8 +44,11 @@ pub struct Job {
 pub trait ClientPool: Send + Sync + 'static {
     /// Assign the given job to the clients.
     /// Returns a list of pair (SocketAddr, Vec<Job>)
+    ///
+    /// When exclude_addresses filters all clients it is ignored.
     async fn assign_jobs(
         &self,
         jobs: Vec<Job>,
+        exclude_addresses: &Option<HashSet<SocketAddr>>,
```

Comment:
It would have been more natural to have `exclude_addresses` be a `HashSet<SocketAddr>`.
When empty we do not exclude any address.

The contract becomes simpler, and so does the code.
As you know an empty HashSet does not allocate so it is not more expensive either.

Also, this implementation does something extremely unexpected if all addresses are discarded.
Here are 3 options:
a) document it
b) remove it entirely if it is not useful. (we just return an error if no node is available)
c) remove it and move that logic to the caller.

I don't like a) because contracts should be as simple as possible. (Complexity should stay within boxes on which we slap on labels with a contract on it.)
b) seems reasonable but I might be missing some information.
c) is ok but might be tricky as it might require information about the pool.

So I suspect c) would lead to something even more complex.
b) is my favorite, and then a) is ok.

Reply:
Ah yeah removing the Option is the best way. It was previously a whitelist, there it made sense.

It is already documented, but on the trait. 
The problem with b is, that the current logic retries if all nodes fail, so this behaviour would be weird, because it depends on how many nodes are assigned to our jobs.

c) would be the cleanest solution since there is no implicit logic, but it adds more complexity.